src/ui.js

```diff
@@ -78,42 +78,29 @@ function handleFileSelect(map, evt) {
 
     modal.show();
 
-    let i = 0;
     let parseFailures = [];
 
-    let loadFile = (file) => {
-        return new Promise(resolve => {
-            let reader = new FileReader();
-            reader.onload = () => resolve(reader.result);
-            reader.readAsText(file, 'UTF-8');
-        }).then(data => {
-            parseGPX(data, (err, track) => {
-                if (err) {
-                    parseFailures.push({name: file.name, error: err});
-                } else {
-                    track.filename = file.name;
-                    tracks.push(track);
-                }
-
-                modal.progress(++i);
-            });
-        });
-    }
+    let openFile = file => new Promise(resolve => {
+        let reader = new FileReader();
+        reader.onload = () => resolve(reader.result);
+        reader.readAsText(file, "UTF-8");
+    });
 
-    Promise.all(files.map(f => loadFile(f)))
-        .then(() => {
-            tracks.forEach(t => {
-                console.log('Adding track:', t.name);
-                map.addTrack(t);
-            });
+    let loadFile = file => openFile(file).then(parseGPX).then(track => {
+        track.filename = file.name;
+        tracks.push(track);
+        console.info('Adding track:', track.name);
+        map.addTrack(track);
+        modal.addSuccess();
+    }, err => {
+        modal.addFailure({name: file.name, error: err});
+    });
```

Comment:
open, parse and add to map, handling failure through the second argument to `.then()`